staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go

```diff
@@ -697,6 +707,73 @@ func (s *Scheme) addObservedVersion(version schema.GroupVersion) {
 	s.observedVersions = append(s.observedVersions, version)
 }
 
+// objectAPILifecycle extracts the APILifecycle from Object interface.
+// returns the APILifecycle, and whether the object has any lifecycle specified.
+func objectAPILifecycle(obj Object) (schema.APILifecycle, bool) {
+	hasLifecycle := false
+	ret := schema.APILifecycle{}
+	introduced, isIntroduced := obj.(apiLifecycleIntroduced)
+	if isIntroduced {
+		major, minor := introduced.APILifecycleIntroduced()
+		ret.IntroducedVersion = version.MajorMinor(uint(major), uint(minor))
+		hasLifecycle = true
+	}
+	removed, isRemoved := obj.(apiLifecycleRemoved)
+	if isRemoved {
+		major, minor := removed.APILifecycleRemoved()
+		ret.RemovedVersion = version.MajorMinor(uint(major), uint(minor))
+		hasLifecycle = true
+	}
+
+	return ret, hasLifecycle
+}
+
+// Object interface generated from "k8s:prerelease-lifecycle-gen:introduced" tags in types.go.
+type apiLifecycleIntroduced interface {
+	APILifecycleIntroduced() (major, minor int)
+}
+
+// Object interface generated from "k8s:prerelease-lifecycle-gen:removed" tags in types.go.
+type apiLifecycleRemoved interface {
+	APILifecycleRemoved() (major, minor int)
+}
+
+// SetGroupVersionLifecycle sets the APILifecycle for the GroupVersion.
+func (s *Scheme) SetGroupVersionLifecycle(gv schema.GroupVersion, lifecycle schema.APILifecycle) {
+	s.gvToAPILifecycle[gv] = lifecycle
+}
+
+// SetResourceLifecycle sets the APILifecycle for the GroupVersionResource,
+// based on Object interfaces generated from "k8s:prerelease-lifecycle-gen:" tags in types.go.
+// Only needed if the GVR lifecycle is different from the GV lifecycle.
+func (s *Scheme) SetResourceLifecycle(gvr schema.GroupVersionResource, obj Object) {
+	lifecycle, hasLifecycle := objectAPILifecycle(obj)
+	if hasLifecycle {
+		s.gvrToAPILifecycle[gvr] = lifecycle
+	}
+}
+
+// GroupVersionLifecycle returns the APILifecycle for the GroupVersion.
+// Returns empty APILifecycle if the GV is not registered, which means the GV is always available.
+func (s *Scheme) GroupVersionLifecycle(gv schema.GroupVersion) schema.APILifecycle {
+	return s.gvToAPILifecycle[gv]
+}
+
+// ResourceLifecycle returns the APILifecycle for the GroupVersionResource.
+// The IntroducedVersion falls back to GV IntroducedVersion if GVR IntroducedVersion is not set.
+// The RemovedVersion falls back to GV RemovedVersion if GVR RemovedVersion is not set.
+func (s *Scheme) ResourceLifecycle(gvr schema.GroupVersionResource) schema.APILifecycle {
```

Comment:
AFAICT this would be the first time we are storing specific "resource" information into the schema (which typically relates to kinds). I see also that the lifecycle information is generated as a member function on the kinds we use for each resource, but the information relates to the resource, not the kind.

Does this break the separation of resources/kinds we have? Does this mean going forward do we intend to establish an explicit (rather than de facto) 1:1 mapping of kinds<->resources?

/cc @jpbetz

Reply:
It doesn't have to be 1:1, there could be multiple resources pointing to the same kind. But the user has to register that.

Reply:
`scheme.SetResourceLifecycle` takes an example object and reads its lifecycle off of member functions. I dont see how you could have two resources with different lifecycles using the same top-level type with such a paradigm. 

This isn't a problem right now because we have a de facto 1:1 mapping of resources<->top-level kinds. I am wondering if doing it this way makes it explicit.

Reply:
This is a really good point @alexzielenski.

Since `+k8s:prerelease-lifecycle-gen:` is already defined for Kinds (it's defined in types.go files, which I think is fair to say is a kind-centric definition), maybe just expose that information directly here, e.g.  "GetPrereleaseLifecycle()"